### ISSUE TITLE
Add `T_A_subseq_isconstant` and `T_B_subseq_isconstant` Parameters to API

### ISSUE THREAD

**Issue by seanlaw** [TDAmeritrade/stumpy#695]
Currently, STUMPY has trouble cleanly differentiating between a "constant subsequence" and a subsequence with a very, very small standard deviation. By adding `T_A_subseq_isconstant` and `T_B_subseq_isconstant` parameters to our API, we can ignore/override the standard deviation completely when computing distances with:

```
if T_B_subseq_isconstant[j] or T_A_subseq_isconstant[i]:
     pearson = 0.5  # Equivalent to a distance of `m`

if T_B_subseq_isconstant[j] and T_A_subseq_isconstant[i]:
    pearson = 1.0  # Equivalent to a distance of `0.0`
```

Doing this would disambiguate between small stddevs and user defined "constant subsequences" and all stddevs can be taken at face value. The only downside is needing to pass around one or two additional vectors but this is a price worth paying as we try to address [other precision-related issues](https://github.com/TDAmeritrade/stumpy/issues/610).

**Comment by NimaSarajpoor** [comment]
@seanlaw 
I remember you mentioned it once before in a PR or issue. Please allow me to think about the things we discussed before and I will share my thoughts with you here :)

**Comment by NimaSarajpoor** [comment]
@seanlaw 
So, I have been thinking about this and I have a few thoughts that I would like to share with you:

(1) **If we want to do this, then we need to add those params to the naive version as well. Right?** I am not sure but I think these settings alone cannot solve the loss of precision unless we treat the subsequences the same in naive version (hence, add those params to naive verison as well) We can try and see how things go.

(2) side note: As a follow up, let's see what we currently do:
```
# in `main` branch

# performant version
stumpy.stump --> uses `core.prepocess_diagonal`, which reset std value to 1.0 if it is below a threshold. 

# naive version 
tests/naive.py::stump --> uses naive z_norm, where std value is reset to 1.0 if it is below a threshold.
```
**It seems we are doing the same thing in both performant and naive. HOWEVER, I think we are not!** If we take a look at function `core.prepocess_diagonal`, we can see that we do ` T_subseq_isconstant = Σ_T < config.STUMPY_STDDEV_THRESHOLD`. Note that we directly set the pearson value to 0 or 0.5, depending on whether  both subseqs are constant or only one of them is constant. This is different than setting `std` to 1.0. I mean... a subseq `S` can have "std" value less than 1.0, but its distance with a non-constant subseq `S'` may not be 0.5. Now, if we set the "std" of `S` to 1, it does not mean that we treat it as constant. So, the naive version can have different value than the performant one



(3) I think the two new `numpy ndarray` params can make the API complicated and it may be used by probably a very few users only. **My main concern** is how a user can set these params? Let's say I have 1M data points. I cannot look at the data to manually choose whether I should treat a subseq as constant or not. I probably do a preprocessing. In my opinion, that preprocessing may look like this:
```
CONSTANT_STD_MAX = 1e-6 # a very small value
s = core.rolling_std(T, m)
is_constant = np.full_like(s, dtype=bool)
is_constant[s < CONSTANT_STD_MAX] = True
```

which is what we currently do in the function `stumpy/core.py::preprocess_diagonal`

Maybe we can add params `T_A_STD_CONSTANT=None` and `T_B_STD_CONSTANT=None` as I think it is more practical. Btw, we already have such variable `config.STUMPY_STDDEV_THRESHOLD` but a user may want to use two different "std" threshold for two different time series `T_A` and `T_B` (which is probably very rare!)

However, we then need to overwrite the config variable `config.STUMPY_STDDEV_THRESHOLD_A(B)`  if the param `T_A(B)_STD_CONSTANT` is not `None`. Because we want to make sure we treat the subsequences as constant in all modules in the same way. Alternatively, we can add these new params to the API of all modules. Then, I, as the user, need to pass `T` and `m` and the same input `config.STUMPY_STDDEV_THRESHOLD_A` in every API. This may not be convenient for the user.


What do you think?

**Comment by seanlaw** [comment]
> I think these settings alone cannot solve the loss of precision

I want to state very clearly that this issue is **not** meant to address any precision issues related to stddev. This issue is **only** for clearly identifying constant subsequences (rather than using stddev) and helping us to know that any future problems involving stddev will be 100% unrelated to constant subsequences. Does that make sense?

>  (1) If we want to do this, then we need to add those params to the naive version as well. Right?

Yes. We'd only specify them in the additional unit tests that we'd add

> (2) Note that we directly set the pearson value to 0 or 0.5, depending on whether both subseqs are constant or only one of them is constant. This is different than setting std to 1.0.

Yes. Setting the stddev to 1.0 is purely to avoid divide-by-zero in `Σ_T_inverse = 1.0 / Σ_T`. Constant subsequences are still treated the same way. I feel like I'm missing your point here? My goal is to disambiguate/differentiate between constant subsequences and subsequences with small stddevs. In other words, when `T_A_subseq_isconstant is None`, the user is choosing to allow for exceptionally small stddevs, which they are then responsible for resolving if those same values cause a divide-by-zero. Basically, we are recusing ourselves from deciding for the user whether a small stddev is a constant subsequence OR just a subsequence with exceptionally small amplitude. Ultimately, we are removing our assumptions/biases and forcing the user to decide.

> I mean... a subseq S can have "std" value less than 1.0, but its distance with a non-constant subseq S' may not be 0.5. Now, if we set the "std" of S to 1, it does not mean that we treat it as constant. So, the naive version can have different value than the performant one

I think we both agree that the presence of a constant subsequence always takes precedent over any stddev value when evaluating distance. So, if a user blindly/accidentally assigns a non-constant subsequence (i.e., a non zero stddev) as being constant (via `T_A_subseq_isconstant`) then we cannot save them (i.e., we assume that the user knows what they are doing here). However we accomplish this, the naive version must match the performant version. I think we are trying to clean up our assumptions here and forcing the user to handle it.

Currently, I believe this is the tricky part:

```
# core.py::preprocess_diagonal
T_subseq_isconstant = Σ_T < config.STUMPY_STDDEV_THRESHOLD
Σ_T[T_subseq_isconstant] = 1.0
Σ_T_inverse = 1.0 / Σ_T
```

I don't know the answer so I am thinking off the top of my head here but I propose that maybe we do something like:

```
# Define a new Python exception class elsewhere
class ConstantSubsequenceError(Exception):
    pass

# In preprocess_diagonal
M_T, Σ_T = compute_mean_std(T, m)

if T_subseq_isconstant is not None:
    if np.isscaler(T_subseq_isconstant):  # eliminate config.STUMPY_STDDEV_THRESHOLD_A(B)
        Σ_T[Σ_T < float(T_subseq_isconstant)] = 1.0
    else:  # T_subseq_isconstant is a numpy array of bool
        Σ_T[T_subseq_isconstant.astype(bool)] = 1.0

with np.errstate(divide='raise'):
    try:
        Σ_T_inverse = 1.0 / Σ_T
    except FloatingPointError:
        msg = "constant subsequence (with zero standard deviation) "
        msg += "was found. We suggest that you..."
        raise ConstantSubsequenceError(msg)    
```

So, if the user does not specify ANY constant subsequences OR a scalar threshold value (this gives the user flexibility to define separate thresholds), then STUMPY will blindly compute the stddev and assume that ALL subsequences are **not** constant, which could potentially lead to a divide-by-zero error. However, we'll catch this error and make suggestions to the user. note that we'll need to determine which "original" time series (`T_A` or `T_B`) contains the constant subsequence so as to inform the user (not done here).

> (3) I think the two new numpy ndarray params can make the API complicated and it may be used by probably a very few users only.

I think we'd create a simple helper function (exposed in our API) to assist with creating the `bool` array and then we'd mention it in the error handling message above.

> However, we then need to overwrite the config variable config.STUMPY_STDDEV_THRESHOLD_A(B) if the param T_A(B)_STD_CONSTANT is not None. Because we want to make sure we treat the subsequences as constant in all modules in the same way. Alternatively, we can add these new params to the API of all modules. Then, I, as the user, need to pass T and m and the same input config.STUMPY_STDDEV_THRESHOLD_A in every API. This may not be convenient for the user.

I think I would eliminate/remove `config.STUMPY_STDDEV_THRESHOLD_A(B)` altogether. Again, I am really focused on constant subsequences here and not solving precision issues regarding stddev.

**Comment by NimaSarajpoor** [comment]
> and helping us to know that any future problems involving stddev will be 100% unrelated to constant subsequences. Does that make sense?

I can see your point now!


> Yes. Setting the stddev to 1.0 is purely to avoid divide-by-zero in `Σ_T_inverse = 1.0 / Σ_T`. Constant subsequences are still treated the same way. I feel like I'm missing your point here? My goal is to disambiguate/differentiate between constant subsequences and subsequences with small/zero stddevs. 

I get it now. Thanks for the clarification!

> In other words, when `T_A_subseq_isconstant is None`, the user is choosing to allow for exceptionally small stddevs, which they are then responsible for resolving if those same values cause a divide-by-zero. Basically, we are recusing ourselves from deciding for the user whether a small stddev is a constant subsequence OR just a subsequence with exceptionally small amplitude. Ultimately, we are removing our assumptions/biases and forcing the user to decide.

I think we need to consider the case `std==0` separately. I try to explain it below:


> Currently, I believe this is the tricky part:
> 
> ```
> # core.py::preprocess_diagonal
> T_subseq_isconstant = Σ_T < config.STUMPY_STDDEV_THRESHOLD
> Σ_T[T_subseq_isconstant] = 1.0
> Σ_T_inverse = 1.0 / Σ_T
> ```
> 
> I don't know the answer so I am thinking off the top of my head here but I propose that maybe we do something like:
> 
> ```
> # Define a new Python exception class elsewhere
> class ConstantSubsequenceError(Exception):
>     pass
> 
> # In preprocess_diagonal
> M_T, Σ_T = compute_mean_std(T, m)
> 
> if T_subseq_isconstant is not None:
>     if np.isscaler(T_subseq_isconstant):  # eliminate config.STUMPY_STDDEV_THRESHOLD_A(B)
>         Σ_T[Σ_T < float(T_subseq_isconstant)] = 1.0
>     else:  # T_subseq_isconstant is a numpy array of bool
>         Σ_T[T_subseq_isconstant.astype(bool)] = 1.0
> 
> with np.errstate(divide='raise'):
>     try:
>         Σ_T_inverse = 1.0 / Σ_T
>     except FloatingPointError:
>         msg = "constant subsequence (with zero standard deviation) "
>         msg += "was found. We suggest that you..."
>         raise ConstantSubsequenceError(msg)    
> ```
> 
> So, if the user does not specify ANY constant subsequences OR a scalar threshold value (this gives the user flexibility to define separate thresholds), then STUMPY will blindly compute the stddev and assume that ALL subsequences are **not** constant, which could potentially lead to a divide-by-zero error, which is fine. However, we'll catch this error and make suggestions to the user.

I do not think we get divide-by-zero error here if the scalar value `T_subseq_isconstant` is non-negative. I mean `Σ_T[Σ_T <= float(T_subseq_isconstant)] = 1.0` replace a `0` with `1.0`. So, even if user sets `T_subseq_isconstant=0`, we are still fine. Right? Does python raise divide-by-zero error when the denominator in not zero but very small?

Also, how about this:
```
 Σ_is_zero = Σ_T == 0
 if T_subseq_isconstant is not None:
     if np.isscaler(T_subseq_isconstant):  
         Σ_T[Σ_T < float(T_subseq_isconstant)] = 1.0
    else:  
         if np.any(T_subseq_isconstant[Σ_is_zero]) == False: # a subseq with std 0 is treated as non-constant
                 raise error
         Σ_T[T_subseq_isconstant.astype(bool)] = 1.0

else:
     T_subseq_isconstant = Σ_is_zero
     Σ_T[T_subseq_isconstant] = 1.0
```

Also, `Σ_T[Σ_T < float(T_subseq_isconstant)] = 1.0` is ALMOST fine. We might get into trouble due to the loss of precision in Σ_T. **I know your focus is to just handle constant subseqs. But I think this particular line makes a connection between the "loss of precision in Σ_T" and `T_subseq_isconstant`.** I assume you want to do something similar in naive. Is that correct? If yes, then due to  loss of precision in rolling std, `Σ_T < float(T_subseq_isconstant)` can result in different arrays in naive and performant if the exact std and rolling std are different. 


> > However, we then need to overwrite the config variable config.STUMPY_STDDEV_THRESHOLD_A(B) if the param T_A(B)_STD_CONSTANT is not None. Because we want to make sure we treat the subsequences as constant in all modules in the same way. Alternatively, we can add these new params to the API of all modules. Then, I, as the user, need to pass T and m and the same input config.STUMPY_STDDEV_THRESHOLD_A in every API. This may not be convenient for the user.
> 
> I think I would eliminate/remove `config.STUMPY_STDDEV_THRESHOLD_A(B)` altogether. Again, I am really focused on constant subsequences here and not solving precision issues regarding stddev.

In this particular part, I was not talking about the precision. My point here is that we should add the new proposed params to almost all modules of stumpy. Right? For instance, let's say I have a time series `T` and I want to calculate matrix profile using `mp=stump(T, m, T_subseq_isconstant=T_subseq_isconstant)`. I set the param `T_subseq_isconstant` because I want to treat some subseqs as constant. Now, if I want to try another module, e.g. `stumpi` or `motifs` just to see what insights I can get from my data, I want those modules to treat those subseqs as constant. So, are you thinking of changing the API of most modules and not just `stump`/ `stumped`/ `gpu_stump`?

**Comment by seanlaw** [comment]
> So, even if user sets T_subseq_isconstant=0, we are still fine. Right? 

Yes, I believe we would still be fine.

> Does python raise divide-by-zero error when the denominator in not zero but very small?

No, it shouldn't. 

> Also, how about this:

Sorry, your proposed code is really hard to follow. Anything with a triple-nested `if/else` smells like a bad idea. I can't seem to follow your point. Additionally,  `np.any(T_subseq_isconstant[Σ_is_zero]) == False` feels like a bad idea as it is changing the input array provided by the user and we should avoid this.

> Also, Σ_T[Σ_T < float(T_subseq_isconstant)] = 1.0 is ALMOST fine. We might get into trouble due to the loss of precision in Σ_T. 

Can you help me understand why it's "almost fine"? I'm not able to get your point. When a subsequence is labeled as `constant`, the stddev (or the inverse) for that subsequence should never matter, right? 

> then due to loss of precision in rolling std, Σ_T < float(T_subseq_isconstant) can result in different arrays in naive and performant if the exact std and rolling std are different.

Okay, then maybe we NEVER allow a user to set `T_subseq_isconstant` as a scalar value?

> In this particular part, I was not talking about the precision. My point here is that we should add the new proposed params to almost all modules of stumpy. Right? 

We would add it everywhere (possibly even in `mstump`)

> So, are you thinking of changing the API of most modules and not just stump/ stumped/ gpu_stump?

Yes!

**Comment by seanlaw** [comment]
Issue #591 is deeply related with this one

**Comment by seanlaw** [comment]
> (1) when T_subseq_isconstant is None, we just set it to the default Σ_T==0. Right? I mean, if user doesn't want to set this param, we need to take care of it ourselves.

I don't think so. I would not `set it to the default Σ_T==0`. I wouldn't change `Σ_T` at all. Let's play out the scenario. If the user does not set `T_subseq_isconstant` (i.e., it is equal to `None`) then two things may happen:

1. None of the Σ_T values are zero and everything is fine and all distances can be computed
2. One or more values in Σ_T are zero and we get a divide-by-zero error that we will catch and then let the user know what the error is (i.e., appears to contain constant subsequences OR subsequences with too small of an amplitude that its stddev cannot be distinguished from that of a constant subsequence) and how to possibly address it by either providing a `T_subseq_isconstant`, injecting some noise into that subsequence, or removing that subsequence from the time series altogether

> (2) when user set T_subseq_isconstant, we need to check if it can cover Σ_T==0 or not. BUT, as you explained, this is not good because we are changing the user's input. So, the error message you provided before is better.

Indeed!

> I think this is safer for passing the unit tests. Also, I assume the user is advanced enough to obtain such array by simply doing T_subseq_isconstant = Σ_T < std_threshold, and then pass T_subseq_isconstant to stumpy.stump

Yes. It puts the burden onto the user to be clear. Certainly, we'll want to add an FAQ section to our documentation (with code) to explain how to handle this situation with a clear example.

**Comment by NimaSarajpoor** [comment]
Cool! 

> 2\. One or more values in Σ_T are zero and we get a divide-by-zero error that we will catch and then let the user know what the error is (i.e., appears to contain constant subsequences OR subsequences with too small of an amplitude that its stddev cannot be distinguished from that of a constant subsequence) and how to possibly address it by either providing a `T_subseq_isconstant`, injecting some noise into that subsequence, or removing that subsequence from the time series altogethe

We should provide some clear examples/guidelines here to help users understand/learn the solutions. 

---

While I was writing this comment, I thought about the following scenario:

Suppose I have time series `T` and I want to try different window sizes in range `(m, m')`. 

```
for window in range(m, m'):
      stumpy.stump(T, window)
```

Now, we have two cases:
* m < m' , in this case, some of the previously-considered constant subseqs may not be constant with higher window
* m > m', then all of the previously-considered constant subseqs remain constant. But, we may get new subseqs that are constant with lower window. 

I am not saying this cannot be resolved. I am just saying the (new) users might get a little bit annoyed by this. One solution is to go with your initial proposal (letting param accept scalar value), and hope we do not encounter loss-of-preciison-related-issue in unit testing.
